packages/react-dom/src/events/__tests__/DOMEventListenerSystem-test-internal.js

```diff
@@ -0,0 +1,865 @@
+          <div ref={divRef}>Click me!</div>
+        </button>
+      );
+    }
+
+    ReactDOM.render(<Test />, container);
+    expect(container.innerHTML).toBe('<button><div>Click me!</div></button>');
+
+    // Clicking the button should trigger the event callback
+    clickEvent = event => {
+      expect(event).toHaveBeenCalledWith(
+        expect.toEqual({
+          eventPhase: 3,
+          type: 'click',
+          currentTarget: buttonRef.current,
+          target: divRef.current,
+        }),
+      );
+    };
+    let divElement = divRef.current;
+    dispatchClickEvent(divElement);
+    clickEvent = jest.fn();
+
+    // Unmounting the container and clicking should not work
+    ReactDOM.render(null, container);
+    dispatchClickEvent(divElement);
+    expect(clickEvent).toBeCalledTimes(0);
+
+    // Re-rendering the container and clicking should work
+    ReactDOM.render(<Test />, container);
+    divElement = divRef.current;
+    dispatchClickEvent(divElement);
+    expect(clickEvent).toBeCalledTimes(1);
+
+    // Clicking the button should also work, but the eventPhase should be 2
+    clickEvent = event => {
+      expect(event).toEqual(
+        expect.objectContaining({
+          eventPhase: 2,
+          type: 'click',
+          currentTarget: buttonRef.current,
+          target: buttonRef.current,
+        }),
+      );
+    };
+    let buttonElement = buttonRef.current;
+    dispatchClickEvent(buttonElement);
+  });
+
+  it('should correctly handle nested listeners #1', () => {
+    const buttonRef = React.createRef();
+    const targetListerner1 = jest.fn();
+    const targetListerner2 = jest.fn();
+    const targetListerner3 = jest.fn();
+    const targetListerner4 = jest.fn();
+
+    function Test() {
+      const listerner1 = ReactDOM.unstable_createListener(
+        'click',
+        targetListerner1,
+        {capture: true},
+      );
+      const listerner2 = ReactDOM.unstable_createListener(
+        'click',
+        targetListerner2,
+        {capture: true},
+      );
+      const listerner3 = ReactDOM.unstable_createListener(
+        'click',
+        targetListerner3,
+      );
+      const listerner4 = ReactDOM.unstable_createListener(
+        'click',
+        targetListerner4,
+      );
+
+      return (
+        <button
+          listeners={[
+            [listerner1],
+            undefined,
+            [listerner2, null, [[listerner3]], listerner4],
+          ]}
+          ref={buttonRef}>
+          Click me!
+        </button>
+      );
+    }
+
+    ReactDOM.render(<Test />, container);
+
+    let buttonElement = buttonRef.current;
+    dispatchClickEvent(buttonElement);
+    expect(targetListerner1).toHaveBeenCalledTimes(1);
+    expect(targetListerner2).toHaveBeenCalledTimes(1);
+    expect(targetListerner3).toHaveBeenCalledTimes(1);
+    expect(targetListerner4).toHaveBeenCalledTimes(1);
+  });
+
+  it('should correctly handle nested listeners #2', () => {
```

Comment:
Can we use a test description that explains the difference between `#1` and `#2`?